### PR TITLE
[restful] Add RESTful API server

### LINKED ISSUE #117: API server required
Require connecting service / model manager / frontend with API server

### PR DESCRIPTION
**Pull request by YuanmingLeee**
Closes #117 

#### Before submitting this PR, please make sure you have:

- [x] run `python -m pytest tests/` and confirmed all checks still pass.
- [x] verified that any code or assets from external sources are properly credited.

### CODE REVIEW

File: modelci/env-backend.env [hunk 1]
```diff
@@ -0,0 +1,3 @@
+BACKEND_CORS_ORIGINS="http://localhost,http://localhost:3000,http://localhost:8080,https://localhost:3000,https://localhost:8080"
+PROJECT_NAME=modelci
+SECRET_KEY=2a6c03b9ca06cd8fc3cf506f0ba924cb735f15918d54758426fd7282366a5e19
```
Comment (HuaizhengZhang) [inline modelci/env-backend.env:3]: should we put here?

File: modelci/persistence/README.md [hunk 2]
```diff
@@ -42,7 +42,7 @@ See test `test/test_model_service.test_register_model`.
 from modelci.persistence.service import ModelService
 
 model_service = ModelService()
-models = model_service.get_models_by_name('ResNet50')
+models = model_service.get_models_by_pk('ResNet50')
```
Comment (HuaizhengZhang) [inline modelci/persistence/README.md:45]: what  is pk？

File: scripts/setup_env.sh [hunk 3]
```diff
@@ -4,4 +4,6 @@
 # shellcheck disable=SC2046
 export $(grep -v '^#' modelci/env-mongodb.env | xargs -d '\r\n')
 
+export $(grep -v '^#' modelci/env-backend.env | xargs -d '\r\n')
```
Comment (HuaizhengZhang) [inline scripts/setup_env.sh:7]: why exports twice？

File: modelci/types/ao/model_ao.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env python
```
Comment (HuaizhengZhang) [inline modelci/types/ao/model_ao.py:1]: what is ao？

File: scripts/setup_env.sh (hunk 3 above)
Comment (YuanmingLeee) [reply scripts/setup_env.sh:7]: different environment, one for mongodb, one for backend.

Comment (HuaizhengZhang) [reply scripts/setup_env.sh:7]: please make comments

File: modelci/persistence/README.md (hunk 2 above)
Comment (YuanmingLeee) [reply modelci/persistence/README.md:45]: primary keys (commonly used in python backend as Django, etc.) We are able to get by one or more of {name, framework, engine, version (to be updated)}.

File: modelci/types/ao/model_ao.py (hunk 4 above)
Comment (YuanmingLeee) [reply modelci/types/ao/model_ao.py:1]: Use python to execute this file, if made executable. Same as `#!/usr/bin/env bash`. 
These lines are generated by IDE.

File: modelci/env-backend.env (hunk 1 above)
Comment (YuanmingLeee) [reply modelci/env-backend.env:3]: The secret key used for production. The RESTful server structure is simplified from FastAPI project template https://github.com/tiangolo/full-stack-fastapi-postgresql

File: modelci/persistence/README.md (hunk 2 above)
Comment (HuaizhengZhang) [reply modelci/persistence/README.md:45]: to be honest, understanding the function from the name easily is what we pursue. 

I prefer the name, id, etc.

File: modelci/types/ao/model_ao.py (hunk 4 above)
Comment (HuaizhengZhang) [reply modelci/types/ao/model_ao.py:1]: ？

Comment (huangyz0918) [reply modelci/types/ao/model_ao.py:1]: What is `ao` in the file name?
@YuanmingLeee

Comment (huangyz0918) [reply modelci/types/ao/model_ao.py:1]: @YuanmingLeee  AO: Active Object?

Comment (YuanmingLeee) [reply modelci/types/ao/model_ao.py:1]: image
AO (application object). my bad, VO (view object) is more appropriate. 

From https://alitech-private.oss-cn-beijing.aliyuncs.com/1561031481870/Java-huashanxinban.pdf?Expires=1592754105&OSSAccessKeyId=LTAIqKGWQyF6Vd3W&Signature=I5xyksQpXbHNgDR%2B3YvDCT8sMyM%3D

File: modelci/persistence/README.md (hunk 2 above)
Comment (YuanmingLeee) [reply modelci/persistence/README.md:45]: in that case, the function name is too long for me `get_models_by_name_framework_engine_version`

Comment (HuaizhengZhang) [reply modelci/persistence/README.md:45]: get_models?

Comment (YuanmingLeee) [reply modelci/persistence/README.md:45]: done

File: scripts/setup_env.sh (hunk 3 above)
Comment (YuanmingLeee) [reply scripts/setup_env.sh:8]: done

**Comment by HuaizhengZhang** [comment]
> You should add a sample request data here @YuanmingLeee

Agree. Please add and then I will review it

**Comment by YuanmingLeee** [comment]
## Get All Model
GET `/api/v1/model/`

QUERY PARAMETERS
name | value
--|--
name(optional) | string (Name)
framework(optional) | string (Framework) Enum: "TENSORFLOW" "PYTORCH"
engine(optional) | string (Engine) Enum: "NONE" "TFS" "TORCHSCRIPT" "ONNX" "TRT" "TVM" "CUSTOMIZED"
version(optional) | integer (Version)

Sample Response
```
[
  {
    "id": "string",
    "name": "string",
    "framework": "TENSORFLOW",
    "engine": "NONE",
    "version": 0,
    "dataset": "string",
    "acc": 0,
    "task": "string",
    "inputs": [
      {
        "shape": [
          0
        ],
        "dtype": "string",
        "name": "string",
        "format": "FORMAT_NONE"
      }
    ],
    "outputs": [
      {
        "shape": [
          0
        ],
        "dtype": "string",
        "name": "string",
        "format": "FORMAT_NONE"
      }
    ],
    "status": "UNKNOWN",
    "create_time": "2020-06-24T09:15:14Z"
  }
]
```

**Comment by YuanmingLeee** [comment]
## Get Model

PATH PARAMETERS

name | value
--|--
id (required) | string (id)

Sample Response
```
{
  "id": "string",
  "name": "string",
  "framework": "TENSORFLOW",
  "engine": "NONE",
  "version": 0,
  "dataset": "string",
  "acc": 0,
  "task": "string",
  "inputs": [
    {
      "shape": [
        0
      ],
      "dtype": "string",
      "name": "string",
      "format": "FORMAT_NONE"
    }
  ],
  "outputs": [
    {
      "shape": [
        0
      ],
      "dtype": "string",
      "name": "string",
      "format": "FORMAT_NONE"
    }
  ],
  "profile_result": {
    "static_result": "string",
    "dynamic_results": [
      {
        "device_id": "string",
        "device_name": "string",
        "batch": 0,
        "memory": {
          "total_memory": 0,
          "memory_usage": 0,
          "utilization": 0
        },
        "latency": {
          "init_latency": {
            "avg": 0,
            "p50": 0,
            "p95": 0,
            "p99": 0
          },
          "preprocess_latency": {
            "avg": 0,
            "p50": 0,
            "p95": 0,
            "p99": 0
          },
          "inference_latency": {
            "avg": 0,
            "p50": 0,
            "p95": 0,
            "p99": 0
          },
          "postprocess_latency": {
            "avg": 0,
            "p50": 0,
            "p95": 0,
            "p99": 0
          }
        },
        "throughput": {
          "batch_formation_throughput": {
            "avg": 0,
            "p50": 0,
            "p95": 0,
            "p99": 0
          },
          "preprocess_throughput": {
            "avg": 0,
            "p50": 0,
            "p95": 0,
            "p99": 0
          },
          "inference_throughput": {
            "avg": 0,
            "p50": 0,
            "p95": 0,
            "p99": 0
          },
          "postprocess_throughput": {
            "avg": 0,
            "p50": 0,
            "p95": 0,
            "p99": 0
          }
        },
        "ip": "string",
        "create_time": "2020-06-24T09:15:15Z"
      }
    ]
  },
  "status": "UNKNOWN",
  "create_time": "2020-06-24T09:15:15Z"
}
```

**Comment by huangyz0918** [comment]
> ## Get All Model
> GET `/api/v1/model/`
> 
> QUERY PARAMETERS
> 
> name	value
> name(optional)	string (Name)
> framework(optional)	string (Framework) Enum: "TENSORFLOW" "PYTORCH"
> engine(optional)	string (Engine) Enum: "NONE" "TFS" "TORCHSCRIPT" "ONNX" "TRT" "TVM" "CUSTOMIZED"
> version(optional)	integer (Version)
> Sample Response
> 
> ```
> [
>   {
>     "id": "string",
>     "name": "string",
>     "framework": "TENSORFLOW",
>     "engine": "NONE",
>     "version": 0,
>     "dataset": "string",
>     "acc": 0,
>     "task": "string",
>     "inputs": [
>       {
>         "shape": [
>           0
>         ],
>         "dtype": "string",
>         "name": "string",
>         "format": "FORMAT_NONE"
>       }
>     ],
>     "outputs": [
>       {
>         "shape": [
>           0
>         ],
>         "dtype": "string",
>         "name": "string",
>         "format": "FORMAT_NONE"
>       }
>     ],
>     "status": "UNKNOWN",
>     "create_time": "2020-06-24T09:15:14Z"
>   }
> ]
> ```

@YuanmingLeee where is the user? I assume the user record can get from mongo db user. 

WeChat3ebb1a1c2737d4c69f98b02b262d7440

**Pull request merged**